### PR TITLE
Data_Engine: Group objects by property values

### LINKED ISSUE #2203: BHoM_Engine: Group objects by property value
#### Description:

Similar to the existing methods of BH.Engine.Data.Query.FilterBy and .OrderBy, it would be really useful to be able to group a list of objects by the value of a given property.

It is already possible to achieve the intended output by iterating over the FilterBy for each unique value of the property (as demonstrated in the attached script & snapshot) but it would be much better to have this functionality in a single method.

[GroupBy.zip](https://github.com/BHoM/BHoM/files/5604045/GroupBy.zip)

image

### PR DESCRIPTION
**Pull request by JohannaOlin1**
### Issues addressed by this PR

Closes #2203 

Adds method to group objects by their property values.

### Test files

[GroupBy_test.zip](https://github.com/BHoM/BHoM_Engine/files/5636805/GroupBy_test.zip)

### CODE REVIEW

File: Data_Engine/Query/Linq/GroupBy.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Groups objects by their property values. Note that not all object properties have groupable values")]
+        [Input("objects", "List of objects to be grouped. All objects in the list should be of similar type")]
+        [Input("propertyName", "The name of the property to group the list of objects by. Note that the property must be able to be grouped - e.g. numbers / text")]
```
Comment (FraserGreenroyd) [inline Data_Engine/Query/Linq/GroupBy.cs:45]:
```suggestion
        [Input("propertyName", "The name of the property to group the list of objects by.")]
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Minor documentation change suggested.

**Review by michaelhoehn — APPROVED** [review]
LGTM

**Review by FraserGreenroyd — APPROVED** [review]
Test script LGTM

**Pull request merged**